File: lib/handlers.js
var commands = require('./commands'),
	fileMap = {};

/**
 * Send data to the client
 * @param socket client's socket
 * @param cmd  command
 * @param data data body
 */
function send(socket, cmd, data) {
	if (data instanceof Buffer) {
		var head = new Buffer(8);
		head[0] = 0xd0;
		head[1] = 0x02;
		head[2] = 0x0f;
		head[3] = commands.out_cmd[cmd];
		head.writeUInt32LE(data.length, 4);
		socket.write(head);
		socket.write(data);
	} else {
		if (typeof data === 'string') {
			send(socket, cmd, new Buffer(data));
		} else {
			console.error('data is not a buffer', data);
		}
	}
}

module.exports = {
	errorHandler: function (arg) {
		var key = arg.key.toString(16);
		if (key.length === 1) {
			key = '0' + key;
		}
		key = '0x' + key;
		console.error('[Error] Unknow key ' + key);
	},

	messageHandler: function (arg) {
		arg.clients.forEach(function (client) {
			send(client, 'message', '[' + arg.sender.remoteAddress + '] ' +
				arg.data.toString());
		});
	},

	openFileHandler: function (arg) {
		var path = arg.data.toString();
		if (!fileMap[path]) {
			// create map and add current sender to the map
			fileMap[path] = [arg.sender];
		} else {
			// just add current sender to the map
			fileMap[path].push(arg.sender);
		}
		console.log(arg.sender.remoteAddress + ' open a file: ' + arg.data.toString());
	},

	closeFileHandler: function (arg) {
		var path = arg.data.toString();
		if (fileMap[path]) {
			for (var i = fileMap[path].length - 1; i >= 0; i--) {
				if (fileMap[path][i] === arg.sender) {
					fileMap[path].splice(i, 1);
				}
			}
		}
		console.log(arg.sender.remoteAddress + ' close a file: ' + arg.data.toString());
	},

	changeSelectionHandler: function (arg) {
		try {
			var o = JSON.parse(arg.data.toString()),
				path = o.path, // file's path
				sels = o.selections; // selections
			if (fileMap[path]) {
				fileMap[path].forEach(function (client) {
					if (client !== arg.sender) {
						// send to other clients
						send(client, 'changeSelection', JSON.stringify({
							client: arg.sender.remoteAddress,
							path: path,
							selections: sels
						}));
					}
				});
			}
			console.log(arg.sender.remoteAddress + ' change selection: ' + o.path);
			console.log(o.selections);
		} catch (err) {
			console.error('changeSelectionHandler', err);
			console.log(arg.data.toString());
		}
	},

	editFileHandler: function (arg) {
		try {
			var o = JSON.parse(arg.data.toString()),
				path = o.path,
				patch = o.patch;
			if (fileMap[path]) {
				fileMap[path].forEach(function (client) {
					if (client !== arg.sender) {
						send(client, 'editFile', JSON.stringify({
							client: arg.sender.remoteAddress,
							path: path,
							patch: patch
						}));
					}
				});
			}
		} catch (err) {
			console.error('editFileHandler', err);
			console.error(arg.data.toString());
		}
	},

	sendMessage: send
};

File: lib/handlers.py
import config
file_map = {}

def error_handler(key, data, pool, sender):
    print('Unknow Key Error: %s' % hex(key))
    print(data)

def login_handler(key, data, pool, sender):
    print('checking user...', data['user'] in config.users, data['pass'] == config.users[data['user']])
    if data['user'] in config.users and data['pass'] == config.users[data['user']]:
        for client in pool:
            if client.is_alive():
                if client is sender:
                    sender.send('message', '[System]: login successed.')
                    sender.user_name = data['user']
                else:
                    client.send('message', '%s connected.' % data['user'])
    else:
        sender.send('disconnect', '[System]: login failed.')
        sender.close()

def message_handler(key, data, pool, sender):
    for thread in pool:
        thread.send('message', '[%s]: %s' % (sender.user_name, data))

def send_client_list(path):
    client_list = list(map(lambda c: c.user_name, file_map[path]))
    for client in file_map[path]:
        client.send('update_client_list', {
            'path': path,
            'client_list': client_list
            })

def open_file_handler(key, data, pool, sender):
    if data in file_map:
        # file not open yet
        file_map[data].append(sender)
    else:
        # file opened
        file_map[data] = [sender]
    send_client_list(data)
    # print('file_map:\n', file_map)

def close_file_handler(key, data, pool, sender):
    if data in file_map:
        file_map[data].remove(sender)
        if len(file_map[data]) is 0:
            del(file_map[data])
        else:
            send_client_list(data)
    
    # print('file_map:\n', file_map)

def change_selection_handler(key, data, pool, sender):
    path = data['path'] # file's path
    sels = data['selections'] # selections
    if path in file_map:
        for client in file_map[path]:
            if client is not sender and client.is_alive():
                client.send('change_selection', {
                    'client': sender.user_name,
                    'path': path,
                    'selections': sels
                    })
    # print('%s change selection: %s' % (sender.address[0], path))

def edit_file_handler(key, data, pool, sender):
    path = data['path']
    patch = data['patch']
    region_dict = data['region_dict']
    if path in file_map:
        for client in file_map[path]:
            if client is not sender and client.is_alive():
                client.send('edit_file', {
                    'client': sender.user_name,
                    'path': path,
                    'patch': patch,
                    'region_dict': region_dict
                    })
    # print('%s edit_file: %s\npatch:\n%s' % (sender.address[0], path, patch))
